### PR TITLE
ddl: add front-end checks for multi-schema change

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/14766

Problem Summary:

### What is changed and how it works?

- Refactor `callHookOnChanged` so that the hook is not called every time.
- Add unique check for changing schema object, including add/drop/alter/modify columns and indexes.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/multi_schema_change.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+func (d *ddl) MultiSchemaChange(ctx sessionctx.Context, ti ast.Ident) error {
+	schema, t, err := d.getSchemaAndTableByIdent(ctx, ti)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	tzName, tzOffset := ddlutil.GetTimeZone(ctx)
+	job := &model.Job{
+		SchemaID:        schema.ID,
+		TableID:         t.Meta().ID,
+		SchemaName:      schema.Name.L,
```
Comment (wjhuang2016) [inline ddl/multi_schema_change.go:37]: TableName?

File: ddl/ddl.go [hunk 2]
```diff
@@ -812,7 +812,11 @@ func (d *ddl) DoDDLJob(ctx sessionctx.Context, job *model.Job) error {
 	}
 }
 
-func (d *ddl) callHookOnChanged(err error) error {
+func (d *ddl) callHookOnChanged(job *model.Job, err error) error {
+	if job.State == model.JobStateNone {
```
Comment (zimulala) [inline ddl/ddl.go:816]: Do we need to add it to the test callback？

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -4159,18 +4170,22 @@ func checkIsDroppableColumn(ctx sessionctx.Context, t table.Table, spec *ast.Alt
...
 	for _, column := range tblInfo.Columns {
 		if !column.Hidden {
 			visibleColumCnt++
 		}
-		if visibleColumCnt > columnCnt {
-			return nil
-		}
 	}
-	return dbterror.ErrTableMustHaveColumns
+	if visibleColumCnt+addCnt > dropCnt {
```
Comment (zimulala) [inline ddl/ddl_api.go:4181]: How about the  added columns are invisible？

File: ddl/multi_schema_change.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+			}
+		}
+		return nil
+	}
+
+	checkIndexes := func(idxNames []model.CIStr, addToModifyIdx bool) error {
+		for _, idxName := range idxNames {
+			name := idxName.L
+			if _, ok := modifyIdx[name]; ok {
+				return dbterror.ErrOperateSameColumn.GenWithStackByArgs(name)
```
Comment (zimulala) [inline ddl/multi_schema_change.go:93]: s/ErrOperateSameColumn/ErrOperateSameIndex?

**Review by wjhuang2016 — APPROVED** [review]

File: ddl/ddl_api.go (hunk 3 above)
Comment (zimulala) [reply ddl/ddl_api.go:4181]: Will invisible columns be added when new expression indexes are added?

File: ddl/multi_schema_change.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+	err = checkVisibleColumnCnt(t, len(info.AddColumns), len(info.DropColumns))
+	if err != nil {
+		return err
+	}
+
+	return checkAddColumnTooManyColumns(len(t.Cols()) + len(info.AddColumns) - len(info.DropColumns))
+}
+
+func checkOperateSameColumn(info *model.MultiSchemaInfo) error {
```
Comment (zimulala) [inline ddl/multi_schema_change.go:73]:
```suggestion
func checkOperateSameColumnAndIndex(info *model.MultiSchemaInfo) error {
```

Comment (tangenta) [reply ddl/multi_schema_change.go]: Updated.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**